pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java

```diff
@@ -1416,18 +1416,22 @@ public CompletableFuture<TopicExistsInfo> checkTopicExists(TopicName topic) {
         return pulsar.getBrokerService().fetchPartitionedTopicMetadataAsync(
                         topic.isPartitioned() ? TopicName.get(topic.getPartitionedTopicName()) : topic)
                 .thenCompose(metadata -> {
+                    // When the topic has metadata:
+                    // - The topic name is non-partitioned, which means that the topic exists.
+                    // - The topic name is partitioned, please check the specific partition.
                     if (metadata.partitions > 0) {
                         if (!topic.isPartitioned()) {
                             return CompletableFuture.completedFuture(
                                     TopicExistsInfo.newPartitionedTopicExists(metadata.partitions));
-                        } else {
-                            if (topic.getPartitionIndex() < metadata.partitions) {
-                                return CompletableFuture.completedFuture(
-                                        TopicExistsInfo.newNonPartitionedTopicExists());
-                            }
+                        }
+                        if (!topic.isPersistent()) {
```

Comment:
it might be useful to make this `} else if (!topic.isPersistent()) {` since the logic assumes that `topic.isPartitioned()`. It works now, but would break if some new condition gets added in the above if block.

Reply:
I guess there's now test coverage for all cases, so it is fine.